Documentation/release-notes/4509.md

```diff
@@ -1,42 +1,131 @@
+
+```cs
+[assembly: UsesLibrary("org.apache.http.legacy", false)]
+```
 
-Xamarin.Android now applies a setting that will make `r8` more lenient
-to keep things closer to the existing `dx` behavior. By automatically
-passing `-ignorewarnings`, `r8` will not fail the build, but still
-emit warnings.
+NuGet package authors who cannot yet update all dependencies to use
+`HttpURLConnection` are encouraged to add this attribute to their projects as
+soon as possible.
 
-To disable this behavior, a new `$(AndroidR8IgnoreWarnings)` MSBuild
-property can be set to `False` in your `.csproj` file:
+This attribute automatically generates the following element under the
+`<application>` node of _AndroidManifest.xml_ for any app that references the
+NuGet package:
 
 ```xml
-<PropertyGroup>
-  <AndroidR8IgnoreWarnings>False</AndroidR8IgnoreWarnings>
-</PropertyGroup>
+<uses-library android:name="org.apache.http.legacy" android:required="false" />
+```
+
+##### Fallback solution
+
+If updated NuGet package versions are not yet available with the attribute, then
+add the same `UsesLibraryAttribute` directly to the application project instead:
+
+```cs
+[assembly: UsesLibrary("org.apache.http.legacy", false)]
 ```
 
-See the [ProGuard manual][proguard] for details about ProGuard rule syntax.
+##### Background information
+
+The _Missing class: org.apache.http.client_ warning indicates that the app has a
+Java dependency on the Apache HTTP client.  Starting in [Android 6.0 Marshmallow
+(API level 23)][android-6-apache-http], that client was deprecated and moved out
+of _android.jar_ into its own _org.apache.http.legacy.jar_ library.  As a
+result, applications that use Apache HTTP client on Android 6.0 or higher will
+only be able to call the APIs successfully if they declare the use of
+org.apache.http.legacy via a [`<uses-library>`][uses-library] element in
+_AndroidManifest.xml_.  Otherwise, the API calls will fail with exceptions
+similar to:
+
+```txt
+Java.Lang.NoClassDefFoundError: Failed resolution of: Lorg/apache/http/impl/client/DefaultHttpClient
+```
+
+#### How to solve _Missing class: java.lang.ClassValue_
+
+```txt
+R8 : warning : Missing class: java.lang.ClassValue
+R8 : error : Compilation can't be completed because some library classes are missing.
+```
+
+##### Solution
+
+Typically, this warning is safe to suppress:
+
+ 1. Add a new text file to the project, named for example _proguard-rules.pro_.
+
+ 2. Set the contents of the file to:
+
+    ```txt
+    -dontwarn java.lang.ClassValue
+    ```
+
+ 3. Edit the _.csproj_ file in a text editor and add `--pg-conf
+    proguard-rules.pro` to the `AndroidR8ExtraArguments` MSBuild property, for
+    example by adding the following lines:
+
+    ```xml
+    <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+      <AndroidR8ExtraArguments>--pg-conf proguard-rules.pro</AndroidR8ExtraArguments>
+    </PropertyGroup>
+    ```
```

Comment:
I used `$(AndroidR8ExtraArguments)` rather than `@(ProguardConfiguration)` because `@(ProguardConfiguration)` items are currently only passed to R8 if code shrinking is enabled.

I thought about that a bit, and as far as I have been able to imagine so far, I think it would work nicely to change the `<R8/>` task slightly so that it supplies `@(ProguardConfiguration)` items to R8 in the multidex-only scenario too. Now that we know that R8 does many or all of the same dependency checks in multidex-only mode as in code shrinking mode, it seems advantageous to pass the `@(ProguardConfiguration)` items in that scenario too. I haven't yet been able to think of any problems with doing that.

I'll submit a PR for that as soon as I get a chance, if no one beats me to it.

Reply:
They could hit this error with `AndroidLinkTool=r8`, too. Should this first mention the normal way to use proguard rules as the common case?